design/src/aws_config.md

```diff
@@ -0,0 +1,161 @@
+# AWS Configuration RFC
+
+> Status: RFC. For an ordered list of proposed changes see: [Proposed changes](#changes-checklist).
+
+An AWS SDK loads configuration from multiple locations. Some of these locations can be loaded synchronously. Some are
+async. Others may actually use AWS services such as STS or SSO.
+
+This document proposes an overhaul to the Config design to facilitate three things:
+
+1. Future-proof: It should be easy to add additional sources of region and credentials, sync and async, from many
+   sources, including code generated services.
+2. Ergonomic: There should be one obvious way to create an AWS service client. Customers should be able to easily
+   customize the client to make common changes. It should lead customers into the pit of success in terms of sharing
+   things that are expensive to create.
+3. Shareable: A config object should be usable to configure multiple AWS services.
+
+## New customer experience
+
+With the new shared config, customers must always take `aws-config` as a dependency if they want access to the automatic
+configuration loading.
+
+In Cargo.toml:
+
+```toml
+[dependencies]
+aws-sdk-dynamo = "0.1"
+aws-config = "0.5"
+```
+
+When creating a client, customers will create a `SharedConfig` struct, then use it to construct the service client.
+The `SharedConfig` struct is defined centrally in `aws-types`. (See [stability](#stability-and-versioning))
+
+```rust
+// pub use of aws_types::SharedConfig
+use aws_config::SharedConfig;
+
+async fn main() {
+    // loading a shared config may make network calls, etc.
+    // with some custom configuration:
+    let config = aws_config::config_loader()
+            .region(Region::new("us-east-1"))
+            .sleep(async_std_sleep)
+            .load().await;
+
+    // with no custom configuration:
+    let config = aws_config::load_config().await;
```

Comment:
I would assume that this _can_ return an error, yeah? no hidden panics?

Reply:
this is a bit of an open question actually. There are definitely no hidden panics here, but errors wouldn't appear until you tried to use the SDK.

Eg. it's not a problem if your profile file doesn't parse properly if you override everything with environment variables.